core/src/main/java/io/parsingdata/metal/expression/value/reference/CurrentIteration.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2013-2016 Netherlands Forensic Institute
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.parsingdata.metal.expression.value.reference;
+
+import static java.math.BigInteger.ONE;
+import static java.math.BigInteger.ZERO;
+
+import static io.parsingdata.metal.Trampoline.complete;
+import static io.parsingdata.metal.Trampoline.intermediate;
+import static io.parsingdata.metal.expression.value.ConstantFactory.createFromNumeric;
+
+import java.math.BigInteger;
+import java.util.Optional;
+
+import io.parsingdata.metal.Trampoline;
+import io.parsingdata.metal.Util;
+import io.parsingdata.metal.data.ImmutableList;
+import io.parsingdata.metal.data.ParseGraph;
+import io.parsingdata.metal.data.ParseItem;
+import io.parsingdata.metal.data.ParseState;
+import io.parsingdata.metal.encoding.Encoding;
+import io.parsingdata.metal.expression.value.Value;
+import io.parsingdata.metal.expression.value.ValueExpression;
+
+/**
+ * A {@link ValueExpression} that represents the current iteration in an
+ * iterable {@link io.parsingdata.metal.token.Token} (e.g. when inside a
+ * {@link io.parsingdata.metal.token.Rep} or
+ * {@link io.parsingdata.metal.token.RepN}).
+ */
+public class CurrentIteration implements ValueExpression {
+
+    @Override
+    public ImmutableList<Optional<Value>> eval(final ParseState parseState, final Encoding encoding) {
+        final ParseGraph currentIterable = findCurrentIterable(parseState.order, ParseGraph.EMPTY).computeResult();
+        if (currentIterable.isEmpty()) { return ImmutableList.create(Optional.empty()); }
+
+        final BigInteger currentIteration = countIterable(currentIterable, ZERO).computeResult();
+        return ImmutableList.create(Optional.of(createFromNumeric(currentIteration, new Encoding())));
+    }
+
+    private Trampoline<ParseGraph> findCurrentIterable(final ParseItem item, final ParseGraph iterableCandidate) {
+        if (!item.isGraph()) { return complete(() -> iterableCandidate); }
```

Comment:
What would happen if I would do something like `rep(tie(any("val"), CURRENT_ITERATION))`? Because when I eval the expression for Tie, there is no value to stop at? Seems like it would blow up?

Reply:
If you mean infinite loop, no, but there will be a null pointer exception, because the head becomes null. Nice catch! Will add `item == null` to this if statement and add tests.

Reply:
Ah, I see your point now. Because a def of size 0 is not added to the parseGraph. However, then the iterable is in fact empty, which will return an empty value and the parsing will fail, see line 50. So it will not blow up.

It is a nice discussion though what you would expect to be "an iteration". Should a parsing attempt, where eventually 0 bytes are parsed, be considered an iteration? E.g. would we expect `rep(def("val", add(CURRENT_ITERATION, con(1))))` to work?

Reply:
Actually, if I read the code as is, I think your first comment does happen (I indeed meant something like a NPE)? Because there is no value yet when you eval CURRENT_ITERATION, and the only way the recursion stops is if it encounters a value? Could you add these samples as a test?

As to the discussion, I would probably expect your example to work, because we are in fact in an iteration? I also think conceptually a 0 bytes parsed should still be considered an iteration, as I feel it is the iteration that happens, not what happens inside the iteration (wow, confusing sentence :smile:).

Reply:
I implemented CURRENT_INDEX differently. Instead of extracting it from ParseGraph, it is kept in the ParseState. I added tests for examples similar to `rep(def("val", add(CURRENT_ITERATION, con(1))))`.

I will need to improve the code a little before I commit.